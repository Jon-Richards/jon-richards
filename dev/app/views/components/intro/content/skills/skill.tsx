import * as React from 'react';
const CSS = require('./skills.scss');

type SkillProps = {
  /** UUID for this particular skill. */
  uuid: string;
  /** The label as it will display to the user. */
  displayLabel: string;
  /** If this is a core skill. */
  isCore: boolean;
};

/**
 * Renders a single skill.
 *
 * @example
 * <Skill
 *      uuid="abc-123"
 *      displayLabel="foo"
 *      value="bar"
 *      isActive={true}
 *      isCore={false}
 *      onClick={(e, value) => {
 *          console.log(value);
 *      }}
 * />
 */
function Skill (props: SkillProps): JSX.Element {
  const { displayLabel, isCore } = props;

  return (
    <span
      className={
        `${CSS['skill']}` +
        `${isCore ? ' ' + CSS['skill--is-core'] : ''}`
      }
    >
      {displayLabel}
    </span>
  );
}

const skillMemo = React.memo<SkillProps>(Skill);

export {skillMemo as Skill, SkillProps};