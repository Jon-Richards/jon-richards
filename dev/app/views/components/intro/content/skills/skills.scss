/**
 * @fileoverview
 * Styling for the skills list in the intro.
 */

@import '~scss-standards';

.root {
  display: flex;
}

.skills {
  display: flex;
  list-style-type: none;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  width: auto;
  height: 100%;
  max-width: 100%;

  @media (min-width: std-get-breakpoint(720)) {
    justify-content: flex-start;
  }
}

.skill {
  display: none;
  margin: 0.2rem 0.2rem;
  padding: 0.4rem 1rem;
  width: auto;
  background: std-get-swatch(1);
  color: std-get-swatch(2, -20);

  @media (min-width: std-get-breakpoint(480)) {
    &--is-core {
      display: flex;
    }
  }

  @media (min-width: std-get-breakpoint(720)) {
    display: flex;
  }
}
