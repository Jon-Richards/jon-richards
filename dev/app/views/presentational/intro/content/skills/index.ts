export { Skills, SkillsProps } from './skills';
