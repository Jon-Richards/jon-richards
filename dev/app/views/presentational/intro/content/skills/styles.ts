import { Breakpoints } from 'Config/styles/breakpoints';

/** Styles for the Skills component. */
export const STYLES = {
  root: `
    display: flex;
  `,
  skills: `
    display: flex;
    list-style-type: none;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    width: auto;
    height: 100%;
    max-width: 100%;

    @media (min-width: ${Breakpoints[720]}) {
      justify-content: flex-start;
    }
  `,
};
