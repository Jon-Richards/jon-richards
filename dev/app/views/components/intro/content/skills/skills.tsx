/**
 * @fileoverview
 * Contains the component that renders a list of skills.
 */

import * as React from 'react';
const CSS = require('./skills.scss');
import { Skill, SkillProps } from './skill';

type SkillsProps = {
  /** Array of skill filters. */
  skills: SkillProps[];
};

/**
 * Renders a filterable list of skills.
 * 
 * __NOTE:__  At the time of this writing, filtering happens automatically based
 * on the isCore property for each skill. 
 *
 * @example
 * <Skills
 *      skills={[
 *          {
 *              displayLabel: 'foo',
 *              value: 'bar',
 *              uuid: 'abc-123',
 *              isCore: true
 *          }
 *      ]}
 * />
 */
function Skills (props: SkillsProps): JSX.Element {
  const { skills } = props;

  return (
    <div className={CSS['root']}>
      <ul className={CSS['skills']}>{mapSkills(skills)}</ul>
    </div>
  );
}

function mapSkills(skills: SkillProps[]): React.ReactFragment {
  if (skills.length) {
    return skills.map(skill => (
      <li key={skill.uuid}>
        <Skill
          displayLabel={skill.displayLabel}
          uuid={skill.uuid}
          isCore={skill.isCore}
        />
      </li>
    ));
  }
  return <></>;
}

const skillsMemo = React.memo<SkillsProps>(Skills);

export { skillsMemo as Skills, SkillsProps };
