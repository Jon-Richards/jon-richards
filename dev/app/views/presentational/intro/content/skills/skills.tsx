/** @jsx jsx */

import * as React from 'react';
import { css, jsx } from '@emotion/core';
import { STYLES } from './styles';
import { Skill, SkillProps } from './skill/skill';

type SkillsProps = {
  /** Array of skill filters. */
  skills: SkillProps[];
};

/**
 * Renders a filterable list of skills.
 * 
 * __NOTE:__  At the time of this writing, filtering happens automatically based
 * on the isCore property for each skill. 
 *
 * @example
 * <Skills
 *      skills={[
 *          {
 *              displayLabel: 'foo',
 *              value: 'bar',
 *              uuid: 'abc-123',
 *              isCore: true
 *          }
 *      ]}
 * />
 */
function Skills (props: SkillsProps): JSX.Element {
  const { skills } = props;

  return (
    <div css={css(STYLES.root)}>
      <ul css={css(STYLES.skills)} data-testid="intro__content__skills__list">
        {mapSkills(skills)}
      </ul>
    </div>
  );
}

function mapSkills(skills: SkillProps[]): React.ReactFragment {
  if (skills.length) {
    return skills.map(skill => (
      <li key={skill.uuid}>
        <Skill
          displayLabel={skill.displayLabel}
          uuid={skill.uuid}
          isCore={skill.isCore}
        />
      </li>
    ));
  }
  return <React.Fragment></React.Fragment>;
}

const skillsMemo = React.memo<SkillsProps>(Skills);

export { skillsMemo as Skills, SkillsProps };
