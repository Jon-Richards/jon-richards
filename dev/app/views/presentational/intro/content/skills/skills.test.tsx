import * as React from 'react';
import { shallow } from 'enzyme';
import { Skills, SkillsProps } from './skills';

const mockSkills: SkillsProps['skills'] = [
  {
    displayLabel: 'Foo',
    isCore: true,
    uuid: '277d793d-1585-4870-9200-acee39522bd4'
  },
  {
    displayLabel: 'Bar',
    isCore: true,
    uuid: '66b58521-fab3-432e-8ab1-63f3041297ec'
  },
  {
    displayLabel: 'Baz',
    isCore: true,
    uuid: '6ff752ab-f9b7-4f25-bcc4-c2bed53ea6d3'
  },
];

describe('The Skills component.', () => {
  it('Should match the snapshot.', () => {
    const test_1 = shallow(
      <Skills skills={mockSkills} />
    );

    expect(test_1).toMatchSnapshot();
  });

  it('Should render according to its props.', () => {
    const test_1 = shallow(
      <Skills skills={mockSkills} />
    );
    const test_2 = shallow(
      <Skills skills={[]} />
    );

    expect(test_1.find('[data-testid="intro__content__skills__list"]')
      .children().length).toBe(3);
    expect(test_2.find('[data-testid="intro__content__skills__list"]')
      .children().length).toBe(0);
  });
});
