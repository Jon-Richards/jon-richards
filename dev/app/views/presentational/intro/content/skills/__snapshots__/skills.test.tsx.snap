// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Skills component. Should match the snapshot. 1`] = `
<li
  key="277d793d-1585-4870-9200-acee39522bd4"
>
  <Memo(Skill)
    displayLabel="Foo"
    isCore={true}
    uuid="277d793d-1585-4870-9200-acee39522bd4"
  />
</li>
<li
  key="66b58521-fab3-432e-8ab1-63f3041297ec"
>
  <Memo(Skill)
    displayLabel="Bar"
    isCore={true}
    uuid="66b58521-fab3-432e-8ab1-63f3041297ec"
  />
</li>
<li
  key="6ff752ab-f9b7-4f25-bcc4-c2bed53ea6d3"
>
  <Memo(Skill)
    displayLabel="Baz"
    isCore={true}
    uuid="6ff752ab-f9b7-4f25-bcc4-c2bed53ea6d3"
  />
</li>
`;
